Automatic conan settings detection issue

## Summmary

There is an issue, when using `conan.cmake` in subdirectory, after `conan_basic_setup` is already invoked in parent project.


Below is the project structure:

```
/test-project
├── CMakeLists.txt
├── conanfile.py
└── sub-project
   ├── CMakeLists.txt
   └── conanfile.py
```

<details><summary>See CMake configure log</summary>
<p>

```log
-- The CXX compiler identification is GNU 9.2.0
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- COMPILE_OPTIONS= | COMPILE_DEFINITIONS= | /test-project
-- Downloading conan.cmake [v0.14] from https://github.com/conan-io/cmake-conan
-- Conan: checking conan executable in path
-- Conan: Found program /usr/bin/conan
-- Conan: Version found Conan version 1.18.4

-- Conan: Automatic detection of conan settings from cmake
-- Conan: Settings= -s;build_type=Release;-s;compiler=gcc;-s;compiler.version=9;-s;compiler.libcxx=libstdc++11
-- Conan executing: conan install /test-project/conanfile.py -s build_type=Release -s compiler=gcc -s compiler.version=9 -s compiler.libcxx=libstdc++11 -g=cmake -if=/test-project/cmake-build-test
Configuration:
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++11
compiler.version=9
os=Linux
os_build=Linux
[options]
[build_requires]
[env]

conanfile.py (alice/0.1.1): Installing package
Requirements
Packages

conanfile.py (alice/0.1.1): Generator cmake created conanbuildinfo.cmake
conanfile.py (alice/0.1.1): Generator txt created conanbuildinfo.txt
conanfile.py (alice/0.1.1): Generated conaninfo.txt
conanfile.py (alice/0.1.1): Generated graphinfo
-- Conan: Loading conanbuildinfo.cmake
-- Conan: Adjusting output directories
-- Conan: Using cmake global configuration
-- Conan: Adjusting default RPATHs Conan policies
-- Conan: Adjusting language standard
-- Current conanbuildinfo.cmake directory: /test-project/cmake-build-test
-- Conan: Compiler GCC>=5, checking major version 9
-- Conan: Checking correct version: 9
-- COMPILE_OPTIONS=$<$<CONFIG:Debug>:>;$<$<CONFIG:Release>:>;$<$<CONFIG:RelWithDebInfo>:>;$<$<CONFIG:MinSizeRel>:> | COMPILE_DEFINITIONS= | /test-project
-- COMPILE_OPTIONS=$<$<CONFIG:Debug>:>;$<$<CONFIG:Release>:>;$<$<CONFIG:RelWithDebInfo>:>;$<$<CONFIG:MinSizeRel>:> | COMPILE_DEFINITIONS= | /test-project/sub-project
-- Downloading conan.cmake [v0.14] from https://github.com/conan-io/cmake-conan
-- Conan: checking conan executable in path
-- Conan: Found program /usr/bin/conan
-- Conan: Version found Conan version 1.18.4

-- Conan: Automatic detection of conan settings from cmake
c++: error: $<$<CONFIG:Debug>:>: No such file or directory
c++: error: $<$<CONFIG:Release>:>: No such file or directory
c++: error: $<$<CONFIG:RelWithDebInfo>:>: No such file or directory
c++: error: $<$<CONFIG:MinSizeRel>:>: No such file or directory
-- Conan: Settings= -s;build_type=Release;-s;compiler=gcc;-s;compiler.version=9;-s;compiler.libcxx=libc++
-- Conan executing: conan install /test-project/sub-project/conanfile.py -s build_type=Release -s compiler=gcc -s compiler.version=9 -s compiler.libcxx=libc++ -g=cmake -if=/test-project/cmake-build-test/sub-project
ERROR: Invalid setting 'libc++' is not a valid 'settings.compiler.libcxx' value.
Possible values are ['libstdc++', 'libstdc++11']
Read "http://docs.conan.io/en/latest/faq/troubleshooting.html#error-invalid-setting"
CMake Error at cmake-build-test/sub-project/conan.cmake:399 (message):
  Conan install failed='1'
Call Stack (most recent call first):
  cmake-build-test/sub-project/conan.cmake:485 (conan_cmake_install)
  sub-project/CMakeLists.txt:43 (conan_cmake_run)


-- Configuring incomplete, errors occurred!
```

</p>
</details>

<details><summary>CMakeLists.txt (same in subdirectory)</summary>
<p>

```cmake
cmake_minimum_required(VERSION 3.8)

project(Alice
    VERSION 0.1.1
    LANGUAGES CXX
)


# DEBUG {{{
    get_directory_property(__co DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR} COMPILE_OPTIONS)
    get_directory_property(__cd DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR} COMPILE_DEFINITIONS)
    message(STATUS "COMPILE_OPTIONS=${__co} | COMPILE_DEFINITIONS=${__cd} | ${CMAKE_CURRENT_SOURCE_DIR}")
    unset(__co)
    unset(__cd)
# }}}

if (CONAN_EXPORTED)
    include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
    conan_basic_setup()
else(CONAN_EXPORTED)
    if(NOT CMAKE_CONAN_VERSION)
        set(CMAKE_CONAN_VERSION 0.14)
    endif(NOT CMAKE_CONAN_VERSION)

    # Download automatically, you can also just copy the conan.cmake file
    if(NOT EXISTS "${CMAKE_CURRENT_BINARY_DIR}/conan.cmake")
        message(STATUS "Downloading conan.cmake [v${CMAKE_CONAN_VERSION}] from https://github.com/conan-io/cmake-conan")
        file(DOWNLOAD "https://raw.githubusercontent.com/conan-io/cmake-conan/v${CMAKE_CONAN_VERSION}/conan.cmake"
            "${CMAKE_CURRENT_BINARY_DIR}/conan.cmake")
    endif()
    include(${CMAKE_CURRENT_BINARY_DIR}/conan.cmake)
    conan_check(REQUIRED)
    conan_cmake_run(CONANFILE conanfile.py BASIC_SETUP)
endif(CONAN_EXPORTED)


# DEBUG {{{
    get_directory_property(__co DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR} COMPILE_OPTIONS)
    get_directory_property(__cd DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR} COMPILE_DEFINITIONS)
    message(STATUS "COMPILE_OPTIONS=${__co} | COMPILE_DEFINITIONS=${__cd} | ${CMAKE_CURRENT_SOURCE_DIR}")
    unset(__co)
    unset(__cd)
# }}}

add_subdirectory(sub-project)
```

</p>
</details>

### Reason

After `conan_basic_setup` there are strange options populated via `COMPILE_OPTIONS` directory property: `COMPILE_OPTIONS=$<$<CONFIG:Debug>:>;$<$<CONFIG:Release>:>;$<$<CONFIG:RelWithDebInfo>:>;$<$<CONFIG:MinSizeRel>:>`

### Solution

Probably, it is issue in Conan CMake generator?
I can propose a quick fix to filter out such options when detecting settings.

--- a/conan.cmake
+++ b/conan.cmake
@@ -33,7 +33,7 @@
 # but it is only necessary on the end-user side. It is not necessary to create conan
 # packages, in fact it shouldn't be use for that. Check the project documentation.
 
-# version: 0.15.0-dev
+# version: 0.15.0
 
 include(CMakeParseArguments)
 

--- a/conan.cmake
+++ b/conan.cmake
@@ -264,9 +264,38 @@ function(conan_cmake_detect_unix_libcxx result)
         endif()
     endforeach()
 
+    # add additional compiler options ala cmRulePlaceholderExpander::ExpandRuleVariable
+    set(EXPAND_CXX_COMPILER ${CMAKE_CXX_COMPILER})
+    if(CMAKE_CXX_COMPILER_ARG1)
+        # CMake splits CXX="foo bar baz" into CMAKE_CXX_COMPILER="foo", CMAKE_CXX_COMPILER_ARG1="bar baz"
+        # without this, ccache, winegcc, or other wrappers might lose all their arguments
+        separate_arguments(SPLIT_CXX_COMPILER_ARG1 NATIVE_COMMAND ${CMAKE_CXX_COMPILER_ARG1})
+        list(APPEND EXPAND_CXX_COMPILER ${SPLIT_CXX_COMPILER_ARG1})
+    endif()
+
+    if(CMAKE_CXX_COMPILE_OPTIONS_TARGET AND CMAKE_CXX_COMPILER_TARGET)
+        # without --target= we may be calling the wrong underlying GCC
+        list(APPEND EXPAND_CXX_COMPILER "${CMAKE_CXX_COMPILE_OPTIONS_TARGET}${CMAKE_CXX_COMPILER_TARGET}")
+    endif()
+
+    if(CMAKE_CXX_COMPILE_OPTIONS_EXTERNAL_TOOLCHAIN AND CMAKE_CXX_COMPILER_EXTERNAL_TOOLCHAIN)
+        list(APPEND EXPAND_CXX_COMPILER "${CMAKE_CXX_COMPILE_OPTIONS_EXTERNAL_TOOLCHAIN}${CMAKE_CXX_COMPILER_EXTERNAL_TOOLCHAIN}")
+    endif()
+
+    if(CMAKE_CXX_COMPILE_OPTIONS_SYSROOT)
+        # without --sysroot= we may find the wrong #include <string>
+        if(CMAKE_SYSROOT_COMPILE)
+            list(APPEND EXPAND_CXX_COMPILER "${CMAKE_CXX_COMPILE_OPTIONS_SYSROOT}${CMAKE_SYSROOT_COMPILE}")
+        elseif(CMAKE_SYSROOT)
+            list(APPEND EXPAND_CXX_COMPILER "${CMAKE_CXX_COMPILE_OPTIONS_SYSROOT}${CMAKE_SYSROOT}")
+        endif()
+    endif()
+
+    separate_arguments(SPLIT_CXX_FLAGS NATIVE_COMMAND ${CMAKE_CXX_FLAGS})
+
     execute_process(
         COMMAND ${CMAKE_COMMAND} -E echo "#include <string>"
-        COMMAND ${CMAKE_CXX_COMPILER} -x c++ ${compile_options} -E -dM -
+        COMMAND ${EXPAND_CXX_COMPILER} ${SPLIT_CXX_FLAGS} -x c++ ${compile_options} -E -dM -
         OUTPUT_VARIABLE string_defines
     )
 

--- a/conan.cmake
+++ b/conan.cmake
@@ -251,9 +251,12 @@ endfunction()
 function(conan_cmake_detect_unix_libcxx result)
     # Take into account any -stdlib in compile options
     get_directory_property(compile_options DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR} COMPILE_OPTIONS)
+    string(GENEX_STRIP "${compile_options}" compile_options)
 
     # Take into account any _GLIBCXX_USE_CXX11_ABI in compile definitions
     get_directory_property(defines DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR} COMPILE_DEFINITIONS)
+    string(GENEX_STRIP "${defines}" defines)
+
     foreach(define ${defines})
         if(define MATCHES "_GLIBCXX_USE_CXX11_ABI")
             if(define MATCHES "^-D")

